asynciterator.ts

```diff
@@ -1621,145 +1621,124 @@ export class MultiTransformIterator<S, D = S> extends TransformIterator<S, D> {
+    let item: T | null;
+    let iterator: AsyncIterator<T> | null;
+    let node = buffer._head;
+    while (node !== null) {
+      if (node.value.readable && (item = node.value.read()) !== null)
+        return item;
+
+      node = node.next;
+      // Note - once #45 is merged it will be more efficient to have
+      // a node.value.done check here and to remove the value when it
+      // is done, and to only call _removeEmptySources when an 'end' event
+      // is called independently of the read
+    }
+
+    while (this._buffer.length < this._maxParallelIterators && (iterator = _sources.read()) !== null) {
+      this._addSource(iterator as any);
+      this._buffer.push(iterator);
+
+      if ((item = iterator.read()) !== null)
+        return item;
     }
+    this._removeEmptySources();
+    this.readable = false;
+    return null;
   }
 
-  // Removes sources that will no longer emit items
   protected _removeEmptySources() {
-    this._sources = this._sources.filter((source, index) => {
-      // Adjust the index of the current source if needed
-      if (source.done && index <= this._currentSource)
-        this._currentSource--;
-      return !source.done;
-    });
-    this._fillBuffer();
-  }
-
-  // Reads items from the next sources
-  protected _read(count: number, done: () => void): void {
-    // Start source loading if needed
-    if (this._pending?.loading === false)
-      this._loadSources();
-
-    // Try to read `count` items
-    let lastCount = 0, item : T | null;
-    while (lastCount !== (lastCount = count)) {
-      // Try every source at least once
-      for (let i = 0; i < this._sources.length && count > 0; i++) {
-        // Pick the next source
-        this._currentSource = (this._currentSource + 1) % this._sources.length;
-        const source = this._sources[this._currentSource];
-        // Attempt to read an item from that source
-        if ((item = source.read()) !== null) {
-          count--;
-          this._push(item);
-        }
+    this._buffer.mutateFilter((source: any) => {
+      if (source.done) {
+        this._removeSource(source);
+        return false;
       }
-    }
+      return true;
+    });
 
-    // Close this iterator if all of its sources have been read
-    if (!this._pending && this._sources.length === 0)
+    // TODO: Fix this up when merging #45
+    if (this._buffer.empty && this._sources.done && this._sourceStarted)
       this.close();
-    done();
+    // Check if we are in a position to continue filling the source buffer
+    else // if (this._sources.readable)
+      this.readable = true;
   }
 
-  protected _end(destroy: boolean = false) {
-    super._end(destroy);
-
+  public close() {
+    this._removeSource(this._sources);
     // Destroy all sources that are still readable
     if (this._destroySources) {
-      for (const source of this._sources)
+      for (const source of this._buffer) {
+        this._removeSource(source);
         source.destroy();
```

Comment:
This one of 2 behaviors I am not happy about here is that we don't also destroy any sources that have not been produced by the `this._sources` iterator. But I can't think of a clean way of doing this here as `this._sources` may have a large amount of overhead to generate the rest of the sources to the end